src/BreakpointBrowser/BreakpointBrowser.class.st

```diff
@@ -0,0 +1,83 @@
+"
+Class {
+	#name : #BreakpointBrowser,
+	#superclass : #Object,
+	#category : #BreakpointBrowser
+}
+
+{ #category : #filtering }
+BreakpointBrowser class >> blacklist [
+	"These methods will not be shown by BreakpointBrowser by default. They contain halts but are meant to contain them (for example, they can be testing that executing self halt does indeed raise a halt)
+	Format: Array of {class name symbol. method name symbol} arrays"
+	^ { 
+	{#ClyClassWithHalts. #methodWithHalts2}.
+	{#ClyClassWithHalts. #methodWithHalts1}.
+	{#ClyClassWithProblemMethods. #methodWithHalt}.
+	{#ClyClassWithProblemMethods. #methodWithHalt2}.
+	{#GTSUnitExampleFailingTest. #testWithHalt}.
+	{#Object. #openDebugger}.
+	{#Object. #openDebugger:}.
+	{#ObjectTest. #b}.
+	{#ObjectTest. #b1}.
+	{#ObjectTest. #testHaltIf}.
+	{#ObjectTest. #testHaltOnce}.
+	{#ObjectTest. #testHaltOnCount}.
+	{#UndefinedObjectTest. #testHaltIfNil}.
+	{#UndefinedObjectTest. #testIfNil}.
+	{#UndefinedObjectTest. #testIfNilIfNotNil}.
+	{#UndefinedObjectTest. #testIfNotNilIfNil}.
+	{#CodeSimulationTests. #methodWithHalt}.
+	{#RBRefactoryTestDataApp. #cruft}.
+	{#RBSmalllintTestObject. #codeCruftLeftInMethods}.
+	{#RBDummyRefactoryTestDataApp. #cruft}.
+	 }
+]
+
+{ #category : #filtering }
+BreakpointBrowser class >> filterOutMethodsFromBlacklist: anOrderedCollectionOfRBMethodDefinition [
+	^ anOrderedCollectionOfRBMethodDefinition select: [ :anRGMethodDef | (self methodIsOnBlacklist: anRGMethodDef) not]
+]
+
+{ #category : #filtering }
+BreakpointBrowser class >> filterOutMethodsFromThisPackage: anOrderedCollectionOfRBMethodDefinition [
+	^ anOrderedCollectionOfRBMethodDefinition select: [ :anRGMethodDef |(anRGMethodDef package realPackage == self package) not ]
+]
+
+{ #category : #filtering }
+BreakpointBrowser class >> methodIsOnBlacklist: anRGMethodDef [
+	^ self blacklist anySatisfy: [:classNameSymbolAndMethodNameSymbol |
+			((anRGMethodDef realClass) = (self environment at: (classNameSymbolAndMethodNameSymbol at: 1) ifAbsent: [^false])) and: 
+			[ anRGMethodDef name = (classNameSymbolAndMethodNameSymbol at: 2) ]
+	].
+]
+
+{ #category : #filtering }
+BreakpointBrowser class >> methodsToDisplay [
+	^ SystemNavigation default allMethodsSelect: [ :aMethod | 
+		aMethod hasBreakpoint or: [
+		aMethod containsHalt]].
+]
+
+{ #category : #filtering }
+BreakpointBrowser class >> open [
+	<script>
+	"Opens a method browser showing all methods containing breakpoints (and its flavours like break once), halts (and its flavours like haltOnce), inactive breakpoints (and flavours) and inactive halts (and flavours). Filter out such methods that belong to the HaltManager package."
+	| b |
+	b := ClyQueryBrowser browseMethods: (self filterOutMethodsFromBlacklist: (self filterOutMethodsFromThisPackage: (self methodsToDisplay))) withTitle: 'Breakpoint Browser'.
+	b switchResultTo: (ClySortedQueryResult using: SortMethodsByPackageFunction new )
```

Comment:
I feel that all this code can be part of current calypso plugin for breakpoints.
This method can be defined on ClyQueryBrowser:
```Smalltalk
ClyQueryBrowser openBreakpoints
```
The filtering logic can be moved to ClyActiveBreakpoints query as parameter "all halts or only real halts" with "only real halts" as default value. Then #openBreakpoints can be defined as:
```Smalltalk
b := ClyQueryBrowser openOn: ClyActiveBreakpoints new.
b switchResultTo: (ClySortedQueryResult using: SortMethodsByPackageFunction new).
```

Reply:
I did that (and added the pragma you mentioned above named: <haltOrBreakpointForTesting>).
https://github.com/dupriezt/Calypso/tree/breakpointQueryImprovement

But there seems to be a caching issue going on. When loading a fresh image and running `ClyQueryBrowser browseBreakpointsExcludingHaltsForTesting.`, it shows 37 methods (expected). When running `ClyQueryBrowser browseBreakpoints.` after it, it still shows 37 methods even though it should show the one method that has the pragma in addition.
When loading a fresh image and running `ClyQueryBrowser browseBreakpoints.` first, it shows the expected 38 methods.

Reply:
Let's discus details here https://github.com/pharo-ide/Calypso/pull/383

Reply:
I explained it in Calypso PR